app/src/main/java/com/epfl/drawyourpath/community/TournamentCreationFragment.kt

```diff
@@ -54,10 +89,99 @@ class TournamentCreationFragment : Fragment(R.layout.fragment_tournament_creatio
+            val id = database.getTournamentUID()
+            if (id == null) {
+                Toast.makeText(
+                    mActivity,
+                    "Can't get a tournament id!",
+                    Toast.LENGTH_LONG,
+                ).show()
+                return@setOnClickListener
+            }
+
+            // ask to store the tournament in the database and handle exceptions
+            storeNewTournament(newTournamentParameters, id, creatorId, database)
+
+            // get back to community fragment without waiting for database
+            replaceFragment<CommunityFragment>()
+        }
+    }
+
+    /**
+     * Helper function that returns a database based on the arguments passed to
+     */
+    private fun getDatabase(): Database {
+        return if (arguments?.getBoolean("USE_WORKING_MOCK_DB", false) == true) {
+            MockDatabase()
+        } else if (arguments?.getBoolean("USE_FAILING_MOCK_DB", false) == true) {
+            MockNonWorkingDatabase()
+        } else {
+            FirebaseDatabase()
+        }
+    }
+
+    private fun getAuth(): Auth {
+        return if (arguments?.getBoolean("USE_WORKING_MOCK_AUTH", false) == true) {
+            MockAuth(forceSigned = true)
+        } else if (arguments?.getBoolean("USE_FAILING_MOCK_AUTH", false) == true) {
+            MockAuth()
+        } else {
+            FirebaseAuth()
+        }
+    }
+
+    /**
+     * Helper function to store the newly created tournament asynchronously, displays toasts
+     * to notify of the success of failure of the operation.
+     *
+     * @param params the TournamentParameters data class containing the choices of the creator
+     * @param id the unique id of the tournament (given by the database)
+     * @param creatorId the id of the user creating the tournament
+     * @param db the database in which we store the tournament
+     */
+    private fun storeNewTournament(
+        params: TournamentParameters,
+        id: String,
+        creatorId: String,
+        db: Database,
+    ) {
+        db.addTournament(
+            Tournament(
+                id = id,
+                name = params.name,
+                description = params.description,
+                creatorId = creatorId,
+                startDate = params.startDate,
+                endDate = params.endDate,
+                visibility = params.visibility,
+            ),
+        ).whenComplete { _, exception ->
+            // Display toasts when complete. Note that it does not complete while offline (but the tournament
+            // is stored when the user gets back online)
+            if (exception != null) {
+                Toast.makeText(
+                    mActivity,
+                    "Operation failed: ${exception.message}",
+                    Toast.LENGTH_LONG,
+                ).show()
+            } else {
+                Toast.makeText(mActivity, "Tournament created!", Toast.LENGTH_LONG).show()

```

Comment:
Could redirect to the tournament list or the main activity maybe?

Reply:
This is part of what happens when the future returned by the database is completed, which can take quite some time. By this time, the last instruction of the `OnCreateTournamentClicked()` which returns to the Community fragment has usually already been called. So it's already returning to the main activity. It works like this so that the user leaves the tournament creation fragment even if the database takes some time to respond.

Reply:
I see, but what if an error occurs? Does the user lose all the data he has set?

Reply:
Yes, but it's not a big deal I think. And it seems that errors very rarely happen with Firebase it seems, it works even if the user is offline.